### PR TITLE
[v0.3] Add ContainerFilesystems to ImageFsInfoResponse

### PR DESCRIPTION
**Pull request by kinarashah**
## Problem 

Kubelet complaining about missing image stats in logs: 
```
E1025 00:48:49.497486    6244 eviction_manager.go:212] "Eviction manager: failed to synchronize" err="eviction manager: failed to get HasDedicatedImageFs: missing image stats: &ImageFsInfoResponse{ImageFilesystems:[]*FilesystemUsage{&FilesystemUsage{Timestamp:1729817329497103360,FsId:&FilesystemIdentifier{Mountpoint:/var/lib/docker,},UsedBytes:&UInt64Value{Value:4951261692,},InodesUsed:&UInt64Value{Value:255210,},},},ContainerFilesystems:[]*FilesystemUsage{},}"
``` 
### Issue 
https://github.com/Mirantis/cri-dockerd/issues/408 

## Solution 
In version v1.31, the feature gate KubeletSeparateDiskGC is enabled, and the ImageFsInfoResponse now includes container filesystem information. https://github.com/kubernetes/kubernetes/pull/120914

Leveraging the current rwLayerSize for each container to determine the total writable layer size. 

Note: I attempted to bump Kubernetes up to v1.31, but that requires significant changes for `k8s.io/cri-api/v1alpha2`. So I decided to stick with v1.29 instead which is the minimum version of `k8s.io/cri-api` that introduces `ContainerFilesystems`. (Looks like this part is already being covered here: https://github.com/Mirantis/cri-dockerd/issues/333)

### CODE REVIEW

**Comment by kinarashah** [comment]
@xinfengliu Could you take a look at this?

File: core/imagefs_linux.go [hunk 1]
```diff
@@ -63,6 +63,15 @@ func (ds *dockerService) imageFsInfo() (*runtimeapi.ImageFsInfoResponse, error)
 	}
 	logrus.Debugf("Total used bytes by docker images: %v", totalImageSize)
 
+	var totalRWLayerSize uint64
+	for cid, _ := range ds.containerStatsCache.stats {
```
Comment (xinfengliu) [inline core/imagefs_linux.go:67]: There's a RW lock for accessing `containerStatsCache`, maybe a better way is to add a new method for `containerStatsCache` in `core/stats.go`?

**Comment by xinfengliu** [comment]
@kinarashah Thank you  for the pull request. 

Could you confirm if the PR will break older versions of `kubernetes`?

Also, We generally create PR against the master branch first, then back-port it to other release branches.

**Comment by kinarashah** [comment]
@xinfengliu Thank you for the quick review and context, I'll test it out with older versions and open a PR against master branch today.

**Comment by kinarashah** [comment]
@xinfengliu Updated the PR. Here's the PR for the `master` branch - https://github.com/Mirantis/cri-dockerd/pull/409. 

Also, confirmed that this doesn't break earlier kubernetes versions because the feature flag is disabled by default. Feature flag is enabled by default for the first time with v1.31 and is still in beta.

**Review by xinfengliu — APPROVED** [review]
LGTM. Thanks!

**Comment by xinfengliu** [comment]
Thank you @kinarashah . I have approved both PRs. 
I'm not the maintainer of this repo, so you may also need other folks such as @nwneisen to review and merge them.

**Review by nwneisen — APPROVED** [review]

**Comment by kinarashah** [comment]
@nwneisen Unit test fixed, and `make test` is now passing locally. Also updated the PR for master branch.

**Comment by kinarashah** [comment]
@nwneisen Thanks for restarting the workflow. I noticed two actions are failing, but they don’t appear to be related to this PR. Could you check if any additional changes are needed from my side?

**Comment by nwneisen** [comment]
@kinarashah They do not look like they are related failures. I will merge with them failing

**Review by nwneisen — APPROVED** [review]

**Comment by xinfengliu** [comment]
@kinarashah and @nwneisen 
Shall we get the corresponding PR #409 in master branch merged too?

**Comment by kinarashah** [comment]
@xinfengliu sorry for the delay, I'll update the PR for master branch today.

**Pull request merged**